File: spec/javascripts/models/iteration.spec.js
describe("iteration", function() {

  beforeEach(function() {
    this.iteration = new Iteration();
  });

  describe("initialize", function() {

    it("should assign stories if passed", function() {
      var stories = [1,2,3];
      var iteration = new Iteration({'stories': stories});
      expect(iteration.get('stories')).toEqual(stories);
    });

  });

  describe("defaults", function() {

    it("should have an empty array of stories", function() {
      expect(this.iteration.get('stories')).toEqual([]);
    });

  });

  describe("points", function() {

    beforeEach(function() {
      var Story = Backbone.Model.extend({name: 'story'});
      this.stories = [
        new Story({estimate: 2, story_type: 'feature'}),
        new Story({estimate: 3, story_type: 'feature'}),
        new Story({estimate: 3, story_type: 'bug'}) // Only features count
                                                    // towards velocity
      ];
      this.iteration.set({stories: this.stories});
    });

    it("should calculate its points", function() {
      expect(this.iteration.points()).toEqual(5);
    });

    it("should return 0 for points if it has no stories", function() {
      this.iteration.unset('stories');
      expect(this.iteration.points()).toEqual(0);
    });

    it("should report how many points it overflows by", function() {
      // Should return 0
      this.iteration.set({'maximum_points':2})
      var pointsStub = sinon.stub(this.iteration, 'points');

      // Should return 0 if the iteration points are less than maximum_points
      pointsStub.returns(1);
      expect(this.iteration.overflowsBy()).toEqual(0);

      // Should return 0 if the iteration points are equal to maximum_points
      pointsStub.returns(2);
      expect(this.iteration.overflowsBy()).toEqual(0);

      // Should return the difference if iteration points are greater than
      // maximum_points
      pointsStub.returns(5);
      expect(this.iteration.overflowsBy()).toEqual(3);
    });

  });

  describe("filling backlog iterations", function() {

    it("should return how many points are available", function() {
      var pointsStub = sinon.stub(this.iteration, "points");
      pointsStub.returns(3);

      this.iteration.set({'maximum_points': 5});
      expect(this.iteration.availablePoints()).toEqual(2);
    });

    it("should always accept chores bugs and releases", function() {
      var stub = sinon.stub();
      var story = {get: stub};

      stub.withArgs('story_type').returns('chore');
      expect(this.iteration.canTakeStory(story)).toBeTruthy();
      stub.withArgs('story_type').returns('bug');
      expect(this.iteration.canTakeStory(story)).toBeTruthy();
      stub.withArgs('story_type').returns('release');
      expect(this.iteration.canTakeStory(story)).toBeTruthy();
    });

    it("should accept a feature if there are enough free points", function() {
      var availablePointsStub = sinon.stub(this.iteration, "availablePoints");
      availablePointsStub.returns(3);
      var pointsStub = sinon.stub(this.iteration, 'points');
      pointsStub.returns(1);

      var stub = sinon.stub();
      var story = {get: stub};

      stub.withArgs('story_type').returns('feature');
      stub.withArgs('estimate').returns(3);

      expect(this.iteration.canTakeStory(story)).toBeTruthy();

      // Story is too big to fit in iteration
      stub.withArgs('estimate').returns(4);
      expect(this.iteration.canTakeStory(story)).toBeFalsy();
    });

    // Each iteration should take at least one feature
    it("should always take at least one feature no matter how big", function() {
      var availablePointsStub = sinon.stub(this.iteration, "availablePoints");
      availablePointsStub.returns(1);

      var stub = sinon.stub();
      var story = {get: stub};
      stub.withArgs('story_type').returns('feature');
      stub.withArgs('estimate').returns(2);

      expect(this.iteration.points()).toEqual(0);
      expect(this.iteration.canTakeStory(story)).toBeTruthy();
    });


  });

});

File: spec/models/iteration_spec.rb
require 'rails_helper'

describe Iteration do
  let(:user) { create(:user) }
  let(:project) { create(:project) }
  let(:service) { IterationService.new(project) }

  subject { Iteration.new(service, 1, 2) }

  before do
    project.users << user
    project.suppress_notifications = true
  end

  describe '#points' do
        it 'should not allow bug/chore story to be estimated' do
      # this is in the beforeEach of iteration_spec.js, that's why we're double checking it here
      expect {
        subject << create(:story, project: project, estimate: 3, story_type: 'bug', state: 'accepted', requested_by: user)
      }.to raise_error(ActiveRecord::RecordInvalid)
    end

    it 'should calculate its points' do
      subject << create(:story, project: project, estimate: 2, story_type: 'feature', requested_by: user)
      subject << create(:story, project: project, estimate: 3, story_type: 'feature', state: 'accepted', requested_by: user)

      expect(subject.points).to eq(5)
    end

    it 'should return 0 for points if it has no stories' do
      expect(subject.points).to eq(0)
    end

    it 'should report how many points it overflows by' do
      # Should return 0 if the iteration points are less than maximum_points
      expect(subject).to receive(:points).and_return(1)
      expect(subject.overflows_by).to eq(0)

      # Should return 0 if the iteration points are equal to maximum_points
      expect(subject).to receive(:points).and_return(2)
      expect(subject.overflows_by).to eq(0)

      # Should return the difference if iteration points are greater than
      # maximum_points
      expect(subject).to receive(:points).and_return(5)
      expect(subject.overflows_by).to eq(3)
    end
  end

  describe 'filling backlog iterations' do
    it 'should return how many points are available' do
      expect(subject).to receive(:maximum_points).and_return(5)
      expect(subject).to receive(:points).and_return(3)
      expect(subject.available_points).to eq(2)
    end

    it 'should always accept chores bugs and releases' do
      story = build(:story)

      %w(chore bug release).each do |story_type|
        story.story_type = story_type
        expect(subject.can_take_story?(story)).to be_truthy
      end
    end

    it 'should not accept anything when isFull is true' do
      story = build(:story)

      subject.instance_variable_set('@is_full', true) # I know, this is ugly

      %w(chore bug release).each do |story_type|
        story.story_type = story_type
        expect(subject.can_take_story?(story)).to be_falsey
      end
    end

    it 'should accept a feature if there are enough free points' do
      allow(subject).to receive(:available_points).and_return(3)
      allow(subject).to receive(:points).and_return(1)

      story = build(:story, story_type: 'feature', estimate: 3)

      expect(subject.can_take_story?(story)).to be_truthy

      expect(story).to receive(:estimate).and_return(4)
      expect(subject.can_take_story?(story)).to be_falsey
    end

    it 'should always take at least one feature no matter how big' do
      allow(subject).to receive(:available_points).and_return(1)
      allow(subject).to receive(:points).and_return(0)

      story = build(:story, story_type: 'feature', estimate: 2)

      expect(subject.can_take_story?(story)).to be_truthy
    end
  end
end
